### PR TITLE
Considering best-effort pods when calculating ready task number

### PR DESCRIPTION
**Pull request by sivanzcw**
#646

### CODE REVIEW

File: pkg/scheduler/framework/session.go [hunk 1]
```diff
@@ -297,6 +297,15 @@ func (ssn *Session) Allocate(task *api.TaskInfo, hostname string) error {
 				return err
 			}
 		}
+	} else {
+		// For best-effort pod, it it not necessary to consider the gang constraint.
+		if task.InitResreq.IsEmpty() {
```
Comment (k82cn) [inline pkg/scheduler/framework/session.go]: what's happen if there not enough resources for the other Pods?

Comment (sivanzcw) [reply pkg/scheduler/framework/session.go]: If there are six pods in a job, and the `minA` is `6`, two of the the six pods are best-effort pods, and the others are non-best-effort pods. The best-effort pods will be dispatched succeed and refreshed to `running` while the non-best-effort pods will be remain `pending` if there are not enough resources to satisfy the resoure requests of the job in the cluster.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 2]
```diff
@@ -231,6 +231,10 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 
 		if ssn.JobReady(job) {
 			stmt.Commit()
+		} else if ssn.JobCandidateReady(job) {
```
Comment (k82cn) [inline pkg/scheduler/actions/allocate/allocate.go]: I think that's not necessary to add a new callback; that should ok to check best-effort pods in `JobReadyFn`.

Comment (sivanzcw) [reply pkg/scheduler/actions/allocate/allocate.go]: If the judgement of the best-effort pod is merged into the `JobReady` fuction, when `JobReady` is satisfied, the dispatch of non-best-effort pod will be started, which will cause the non-best-effort pod of the job to be bound to the node first. But at this time, the gang constraint of the job is not satisfied. Maybe the remaining best-effort pods under the job can not be successfully scheduled during the `backfill` action.

Comment (k82cn) [reply pkg/scheduler/actions/allocate/allocate.go]: IMO, we should dispatch all pods of job in allocate action, including best-effort pod :)

File: pkg/scheduler/api/job_info.go [hunk 3]
```diff
@@ -343,17 +343,19 @@ func (ji *JobInfo) FitError() string {
...
-	occupid := 0
-	for status, tasks := range ji.TaskStatusIndex {
-		if AllocatedStatus(status) ||
-			status == Succeeded {
-			occupid = occupid + len(tasks)
+	var occupied int32
+	for _, task := range ji.Tasks {
+		if task.InitResreq.IsEmpty() {
+			occupied++
+		} else if AllocatedStatus(task.Status) ||
```
Comment (k82cn) [inline pkg/scheduler/api/job_info.go]: if so, we're going to go through all pending tasks for this. Use `TaskStatusIndex` to check allocated tasks, and find best-effort pods in pending status.

**Pull request merged**